In [3]:
import nltk, os

In [4]:
s = 'a man in the house' # untokenized string
t = ['a', 'man', 'in', 'the', 'house'] # tokenized seqeuence of words as a list

s 안에 하나의 긴 스트링. 밑은 띄어쓰기 별로 끊어준 것(스플릿으로)

In [5]:
a = nltk.Text(t)
a

<Text: a man in the house...>

nltk에 포함된 text라는 함수 쓰기.

In [6]:
nltk.Text(s.split())

<Text: a man in the house...>

s.split 대신에 위에서 만들어 놓은 t를 넣어도 됨.

In [7]:
b = nltk.Text(s.split())
type(b)

nltk.text.Text

In [9]:
raw = open(os.getcwd()+r'/06_01.txt', encoding = 'utf8').read()
raw

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\samsung\\Class/06_01.txt'

+는 스트링 두개를 합치는 것. 디렉토리 처음부터 끝까지를 다 열어라. 뭐를 가지고 여냐면 utf8로. 그 파일을 열어서 .read 하면 됨. raw에 텍스트가 들어와 있을 것

In [18]:
print(type(raw))
print(len(raw))
print(raw[:75])

<class 'str'>
1154507
The Project Gutenberg EBook of Crime and Punishment, by Fyodor Dostoevsky




타입, 총 길이, 제일 처음부터 75번째까지.

In [19]:
tokens = nltk.word_tokenize(raw)
print(type(tokens))
print(len(tokens))
print(tokens[:10])

<class 'list'>
257726
['The', 'Project', 'Gutenberg', 'EBook', 'of', 'Crime', 'and', 'Punishment', ',', 'by']


nltk 라이브러리 안에 word_tokenize라는 함수가 있다. (raw): raw한 스트링을 담는 것.
타입은 리스트. 스플릿으로 받는 것과 같은 결과. 저 숫자는 단어의 개수. 토크나이즈해서 띄어쓰기 기준으로 끊었으니까.
첫번째에서 열번째까지 보기.위의 결과와 같은 걸 알 수 있음. 약간의 차이가 있다면 ;같은 거를 한 단어 처리해서 끊었던 걸 word_tokenize했기때문에.

In [20]:
text = nltk.Text(tokens)
print(type(text))
print(text[:10])
text.collocations()


<class 'nltk.text.Text'>
['The', 'Project', 'Gutenberg', 'EBook', 'of', 'Crime', 'and', 'Punishment', ',', 'by']
Katerina Ivanovna; Pyotr Petrovitch; Pulcheria Alexandrovna; Avdotya
Romanovna; Rodion Romanovitch; Marfa Petrovna; Sofya Semyonovna; old
woman; Project Gutenberg-tm; Porfiry Petrovitch; Amalia Ivanovna;
great deal; young man; Nikodim Fomitch; Ilya Petrovitch; Project
Gutenberg; Andrey Semyonovitch; Hay Market; Dmitri Prokofitch; Good
heavens


.Text 함수의 역할을 볼 수 있음. 일단 타입은 아까 나온 것(b의 타입)과 같음.
nltk 전용 variable로 만듦. 그 텍스트의 열번째까지를 보면 아웃풋이 위와 같다.
nltk 전용을 만들면 nltk 안에 있는 다양한 함수를 쓸 수 있어서 좋음. .collocation이라는 함수를 쓴 아웃풋이 Katerina부터.
두 단어씩 페어로 제시. 앞 단어가 있으면 뒤의 단어가 같이 나올 확률이 높다는 것. 대부분 사람의 이름. 계속 반복이 되는 페어를 제시.
young man같은 것도 소설 속에서 자주 발견이 되었음을 보여줌. 

In [21]:
text.concordance('great', 79, 25)


Displaying 25 of 162 matches:
and were more frequent in periods of great strain . In 1859 he was allowed to r
ndency and of late she had read with great interest a book she got through Mr. 
 the bosom of her family ... . And a great deal more ... . Quite excusable , si
that you had heard that Dounia had a great deal to put up with in the Svidrigra
g will she has . Dounia can endure a great deal and even in the most difficult 
 that letter she reproached him with great heat and indignation for the basenes
putation ; they had seen and known a great deal more than Mr. Svidrigailov had 
n other people ’ s . In my opinion a great deal , a very great deal of all this
 In my opinion a great deal , a very great deal of all this was unnecessary ; b
 . He is a very busy man and is in a great hurry to get to Petersburg , so that
 me that , though he is not a man of great education , he is clever and seems t
 very well . Of course , there is no great love either on his side , or on hers
tted the m

text를 위에서 만들었으니까 거기에 .concordance 함수. great가 보이는 모든 라인을 spotting을 하는데, 그 문맥을 파악하기 위해 great을 중심으로 총 79개의 character를 봄. 이 숫자를 줄이면 세로로 좁아짐.  25는 처음부터 스물다섯번째의 샘플. 5로 줄이면 5개의 great만.

In [22]:
raw = nltk.corpus.gutenberg.raw('melville-moby_dick.txt')
fdist = nltk.FreqDist(raw)
print(fdist.most_common(5))


fdist = nltk.FreqDist(ch.lower() for ch in raw if ch.isalpha())
print(fdist.most_common(5))
[char for (char, count) in fdist.most_common(5)]

[(' ', 198098), ('e', 115855), ('t', 85539), ('a', 75266), ('o', 68338)]
[('e', 117092), ('t', 87996), ('a', 77916), ('o', 69326), ('n', 65617)]


['e', 't', 'a', 'o', 'n']

nltk 라이브러리는 함수뿐만 아니라 그냥 데이터를 갖다 놓기도 함. 예제 데이터 등. 아나콘다를 깔면서 nltk가 깔렸을 텐데, 아마 멜빌의 모비딕 포함 다양한 텍스트가 안에 있을 것. raw에 텍스트 스트링이 들어 있고, print(raw[:200])하면 처음부터 200까지. 아웃풋 내용은 제일 common한 게 스페이스였고, 그 횟수는 198098이었다. 이하 순위대로. 상위 5개까지. 
most_common은 그중에서 가장 많이 나온 게 무엇인지 알려줌.

control + slash = command / uncommand

ch.lower() for ch in raw if ch.isalpha: .isalpha는 이게 알파벳이냐 하는 것. 알파벳이면 그걸 lower해서 list로 만들어라. 숫자나 문장부호 같은거 있으면 빼고 알파벳만 포함. 대문자는 소문자화. ch는 캐릭터.
--> 이 내용을 가지고 FreqDist한다.

그 결과는 문장부호와 대문자 없이 나옴. 어떤 캐릭터가, 어떤 철자가 가장 많이 나오는구나 알 수 있음. e가 가장 많이 나오고, 모음이 많이 나오는구나 유추할 수 있음.
char for~: 위에 결과는 튜플. 두 개의 정보가 들어 있음. (char, count) 두 개의 정보를 input 했으니까. 총 몇 번 도는가?--> 다섯번. 다섯번을 도는 데 char, 즉 캐릭터만 받아라. 그러면 결과는 리스트. 

In [23]:
nltk.corpus.words.words('en')[-20:-1]


['zymosis',
 'zymosterol',
 'zymosthenic',
 'zymotechnic',
 'zymotechnical',
 'zymotechnics',
 'zymotechny',
 'zymotic',
 'zymotically',
 'zymotize',
 'zymotoxic',
 'zymurgy',
 'Zyrenian',
 'Zyrian',
 'Zyryan',
 'zythem',
 'Zythia',
 'zythum',
 'Zyzomys']

In [24]:
len(nltk.corpus.words.words('en'))

235886